tao_compiler/mlir/disc/transforms/fusion_utils.h

```diff
@@ -227,88 +236,124 @@ class FusionPattern {
-    return getValue() == rhs.getValue();
-  }
+  // Returns the fusion kind of the fusion pattern.
+  FusionType getFusionType() { return fusion_type_; }
 
- private:
-  Value value_;
-};
+  // Returns the fusion kind of the fusion pattern.
+  StringRef getFusionTypeStr() { return fusionTypeToString(fusion_type_); }
 
-bool operator<(const ValueWrapper& lhs, const ValueWrapper& rhs);
+  // Sets the fusion type to the the type provided.
+  void setFusionType(FusionType type) { fusion_type_ = type; }
 
-// This is a simple shape constraint analysis, which is used to
-// guide fusion decision (e.g. we only fuse shape-compatible ops).
-//
-// Currently, We only consider shape equality and same-number-elements equality
-// propagation based on the shape constraint traits of elementwise ops (assuming
-// that implicit shape broadcast is forbidden).
-class ShapeConstraintAnalysis {
- public:
-  explicit ShapeConstraintAnalysis(const SmallVectorImpl<Operation*>& op_list) {
-    PropagateEquality(op_list);
-  }
+  // Returns true if this a fusible fusion pattern.
+  bool isFusible() { return getFusionType() != FusionType::kNone; }
+
+  // Returns true if this fusion pattern is a kLoop fusion.
+  bool isKLoopFusion() { return getFusionType() == FusionType::kLoop; }
 
-  // Returns true if `lhs` and `rhs` are supposed to have same shape.
-  bool HasSameShape(Value lhs, Value rhs) {
-    return same_shape_impl_.isEquivalent(ValueWrapper(lhs), ValueWrapper(rhs));
+  // Returns true if this fusion pattern is a kInput fusion.
+  bool isKInputFusion() {
+    return (getFusionType() == FusionType::kRowReduction ||
+            getFusionType() == FusionType::kColReduction);
   }
 
-  // Returns true if `lhs` and `rhs` are supposed to have same number of
-  // elements.
-  bool HasSameNumElements(Value lhs, Value rhs) {
-    return same_num_elements_impl_.isEquivalent(ValueWrapper(lhs),
-                                                ValueWrapper(rhs));
+  // Returns true if the fusion type is stitch fusion.
+  bool isStitchFusion() { return getFusionType() == FusionType::kStitch; }
+
+  // Merges two fusion patterns and returns the merged pattern. The original
+  // pattern remains unmodified. The new merged pattern is uninitialized.
+  FusionPattern mergeWithoutInit(FusionPattern& other);
+
+  DenseMap<Value, TileInfo>& getTilePlan() { return tile_plan_; }
+  void setTilePlan(const DenseMap<Value, TileInfo>& tile_plan) {
+    tile_plan_ = tile_plan;
   }
 
-  Value GetLeaderValueWithSameShape(Value val) const {
-    if (same_shape_impl_.findLeader(ValueWrapper(val)) ==
-        same_shape_impl_.member_end()) {
-      return nullptr;
-    }
-    return same_shape_impl_.getLeaderValue(ValueWrapper(val)).getValue();
+  SmallVector<Operation*, 4>& getSubRootOps() { return sub_root_ops_; }
+
+  void setSubRootOps(const SmallVectorImpl<Operation*>& sub_root_ops) {
+    sub_root_ops_.insert(sub_root_ops_.begin(), sub_root_ops.begin(),
```

Comment:
Maybe use appendSubRootOps as name is more accurate?

Reply:
Good suggestion. I will update it.